lib/schemes/oauth2.js

```diff
@@ -198,4 +202,85 @@ export default class Oauth2Scheme {
 
     return true // True means a redirect happened
   }
+
+  // ---------------------------------------------------------------
+  // Watch axios requests for token expiration
+  // Refresh tokens if token has expired
+  // ---------------------------------------------------------------
+
+  initializeTokenRefreshOnRequest () {
+    const { $axios } = this.$auth.ctx.app
+    let isRefreshing = false
+
+    $axios.onRequest(async config => {
+      let token = this.$auth.getToken(this.name)
+      let refreshToken = this.$auth.getRefreshToken(this.name)
+
+      // Token or "refresh token" does not exists
+      if (!token || !refreshToken || !token.length || !refreshToken.length) {
+        return config
+      }
+
+      // If already trying to refresh token, do not try again
+      if (isRefreshing) {
+        return config
+      }
+
+      // Time variables
+      let tokenExpiresAt = jwtDecode(token).exp * 1000
+      let refreshTokenExpiresAt = jwtDecode(refreshToken).exp * 1000
+      const now = Date.now()
+
+      // Give us some slack to help the token from expiring between validation and usage
+      const timeSlackMillis = 500
+      tokenExpiresAt -= timeSlackMillis
+      refreshTokenExpiresAt -= timeSlackMillis
+
+
+      // Return if token has not expired
+      if (now < tokenExpiresAt) {
+        return config
+      }
+      // "Refresh token" has also expired. There is no way to refresh. Force logout.
+      if (now > refreshTokenExpiresAt) {
+        this.logout()
+
+        if(process.client) {
+          // Explicitly redirect to the signed-out page.
+          // We don't want to redirect a user with an expired token on page reload/first navigation.
+          // WatchLoggedIn in auth.js->mounted redirects to logout when the state changes,
+          // but it only works in the client
+          this.$auth.redirect('logout')
+        }
+        // Stop the request from happening. The original caller must catch ExpiredSessionErrors
+        throw new ExpiredSessionError()
+      }
+
+      // Try to refresh token before processing current request
+      isRefreshing = true
+
+      return $axios.post(this.options.access_token_endpoint,
+        encodeQuery({
```

Comment:
I think axios itself should handle encoding of query (by passing query param)

Reply:
I had a look at the [axios docs](https://github.com/axios/axios#using-applicationx-www-form-urlencoded-format), but it looks like they suggest using the qs lib to encode it... So I think we must encode ourselves.. It tried the following, but it did not work. 
```js
const data = await this.$auth.request({
  method: 'post',
  headers: { 'content-type': 'application/x-www-form-urlencoded' },
  url: this.options.access_token_endpoint,
  baseURL: process.server ? undefined : false,
  data: {
    code: parsedQuery.code,
    client_id: this.options.client_id,
    redirect_uri: this._redirectURI,
    response_type: this.options.response_type,
    audience: this.options.audience,
    grant_type: this.options.grant_type
  }
})
```
It resulted in the correct content-type, but the form data ended up being sent like json, and the server returned 400:
`{"code":"fancyauthcode","client_id":"myId","redirect_uri":"http://localhost:3000/callback","response_type":"code","grant_type":"authorization_code"}`

Reply:
Yes. But it highly depends on backend implementation. At least the ones i've seen, include this in URL query not sending as form-data. Some may accept JSON. That should be more generic.

Reply:
But we have not been sending the data as query parameters before? It has always been sent as an encoded query string in the request body, right? What do you suggest we do? I still don't know how to make this work without using encodeQuery.